### DISCUSSION TITLE
Does TUnit support constructor parameters?

### DISCUSSION THREAD

**Discussion by tom-englert** [thomhurst/TUnit#589] [Q&A] [answered]
We are using NUnit with heavy use of constructor parameters - this way we can easily run all tests against all supported database types.

Is this supported also by TUnit?

**Comment by thomhurst** [comment]
Yes, see here: https://thomhurst.github.io/TUnit/docs/tutorial-basics/data-source-driven-tests

**Reply by thomhurst** [reply]
The arguments, methoddatasource and ClassDataSource attributes work on both methods and class constructors

**Reply by thomhurst** [reply]
@tom-englert does this help?

**Comment by tom-englert** [comment]
That's not what I need, it's not only about the lifetime of a data source, it's about data source on class level that automatically applies to all tests.
This is our use case:
```c#
public enum TestDbEngine
{
    Mongo,
    EFCorePostgres
}

[TestFixture(TestDbEngine.Mongo)]
[TestFixture(TestDbEngine.EFCorePostgres)]
public class TestClass(TestDbEngine dbEngine)
{
    [OneTimeSetUp]
    public override async Task OneTimeSetUp()
    {
        _dbContainer = dbEngine switch
        {
            TestDbEngine.EFCorePostgres => new PostgreSqlBuilder()
                .WithImage(TestImages.PostgreSql)
                .Build(),
            TestDbEngine.Mongo => new MongoDbBuilder()
                .WithImage(TestImages.MongoDb)
                .Build()
        };

	await _dbContainer.StartAsync();
    }
	
    [TestCase(0)]
    [TestCase(1)]
    public void Test(int testCase)
    {
        // Some test
    }

    ... and many more tests....
```

This will run a matrix of tests with all their test cases against all databases

**Comment by tom-englert** [comment]
```c#
[AttributeUsage(AttributeTargets.Method | AttributeTargets.Class, AllowMultiple = true)]
public sealed class ArgumentsAttribute : TUnitAttribute
{
 ```

Looking at the attribute targets, it should be possible, right?

**Comment by tom-englert** [comment]
Technically works fine:
```c#
[Arguments(1)]
[Arguments(2)]
public class TestClass(int classParameter)
{
    [Test]
    [Arguments(1)]
    [Arguments(2)]
    public async Task TestMethod(int testParameter)
    {
        Console.WriteLine($"Class parameter: {classParameter}, test parameter: {testParameter}");
        
        await Assert.That(1).IsEqualTo(1);
    }
```
However in the test explorer class parameters can't be distinguished:
![image](https://github.com/user-attachments/assets/c2f7172e-2503-45cd-9a44-0254ad390bf1)

**Reply by tom-englert** [reply]
@thomhurst is this an issue in the VS test explorer, or something that can be fixed in TUnit?

**Reply by thomhurst** [reply]
It's a limitation of both testing platform currently, and VS test explorer.

There's discussions here : https://github.com/microsoft/testfx/issues/3371

I'd add in you'd like this feature as I don't think it's got much traction behind it ATM.

**Comment by tom-englert** [comment]
> I've tried setting the TypeName in TestMethodIdentifierProperty to more of a display name. That sort of works, but doesn't give you the same level of organisation of being able to group by class WITHOUT params and then WITH.

If there is no movement in the test API, that would be better than nothing (maybe configurable by attribute?)

**Reply by thomhurst** [reply, marked as answer]
@tom-englert This should work in v0.1.818 now if you want to give it a try

**Comment by tom-englert** [comment]
Yes, that's much better now:
![image](https://github.com/user-attachments/assets/69717472-f624-47a7-83ac-5b423a3ba189)

**Comment by tom-englert** [comment]
@thomhurst btw. any idea how to show the test output (Console or TestContext.Current.OutputWriter) in VS Test explorer?

**Reply by thomhurst** [reply]
I don't believe VS have built it yet as it's still preview and missing some features.

I've done what I need to on my end so hopefully in an update soon it'll just start showing.
